src/engine/Math/Random.ts

```diff
@@ -0,0 +1,253 @@
+
+        this._index = 0;
+      }
+
+      /**
+       * Return next 32 bit integer number in sequence
+       */
+      public nextInt(): number {
+         if (this._index >= this._n) {
+            this._twist();            
+         }
+
+         var y = this._mt[this._index++];
+
+         y ^= y >>> this._u;
+         y ^= ((y << this._s) & this._b);
+         y ^= ((y << this._t) & this._c);
+         y ^= (y >>> this._l);
+ 
+         return y >>> 0;
+      }
+
+      /**
+       * Return a random floating point number between [0, 1)
+       */
+      public next(): number {
+         return this.nextInt() * (1.0 / 4294967296.0); // divided by 2^32
+      }
+
+      /**
+       * Return a random floating point in range [min, max) min is included, max is not included
+       */
+      public floating(min: number, max: number): number {
+         return (max - min) * this.next() + min;
+      }
+
+      /**
+       * Return a random integer in range [min, max] min is included, max is included.
+       * Implemented with rejection sampling, see https://medium.com/@betable/tifu-by-using-math-random-f1c308c4fd9d#.i13tdiu5a
+       */
+      public integer(min: number, max: number): number {
+         return Math.floor((max - min + 1) * this.next() + min);
+      }
+
+      /**
+       * Returns true or false randomly with 50/50 odds by default. 
+       * By default the likelihood of returning a true is .5 (50%).
+       * @param likelihood takes values between [0, 1]
+       */
+      public bool(likelihood: number = .5): boolean {
+         return this.next() <= likelihood;
+      }
+
+      /**
+       * Returns one element from an array at random
+       */
+      public pickOne<T>(array: Array<T>): T {
+         return array[this.integer(0, array.length - 1)];
+      }
+
+      /**
+       * Returns a new array randomly picking elements in the original (not reused)
+       * @param numPicks must be less than or equal to the number of elements in the array.
+       */
+      public pickSet<T>(array: Array<T>, numPicks: number): Array<T> {
+         if (numPicks > array.length || numPicks < 0) {
+            throw new Error('Invalid number of elements to pick, must pick a value 0 < n <= length');
+         }
+         if (numPicks === array.length) {
+            return array;
+         }
+
+         var result: Array<T> = new Array<T>(numPicks);
+         var currentPick = 0;
+         var tempArray = array.slice(0);
+         while (currentPick < numPicks) {
+            var index = this.integer(0, tempArray.length - 1);
+            result[currentPick++] = tempArray[index];
+            tempArray.splice(index, 1);
+         }
+
+         return result;
+      }
+
+      /**
+       * Returns a new array random picking elements from the original allowing duplicates
+       * @param numPicks can be any positive number
+       */
+      public pickSetWithDuplicates<T>(array: Array<T>, numPicks): Array<T> {
+         if (numPicks < 0) {
```

Comment:
This is actually okay, if there are duplicates allowed. Since you can now have an array larger than the source if you allow dups